test/extensions/filters/listener/common/fuzz/listener_filter_fuzzer.cc

```diff
@@ -79,21 +87,33 @@ FuzzedHeader::FuzzedHeader(const test::extensions::filters::listener::FilterFuzz
 
   for (int i = 0; i < nreads_; i++) {
     data_.insert(data_.end(), input.data(i).begin(), input.data(i).end());
-    indices_.push_back(data_.size());
+    indices_.push_back(data_.size() - 1);
   }
 }
 
-Api::SysCallSizeResult FuzzedHeader::next(void* buffer, size_t length) {
-  if (done()) {           // End of stream reached
-    nread_ = nreads_ - 1; // Decrement to avoid out-of-range for last recv() call
+void FuzzedHeader::next() {
+  if (!done()) {
+    nread_++;
   }
-  memcpy(buffer, data_.data(), std::min(indices_[nread_], length));
-  return Api::SysCallSizeResult{static_cast<ssize_t>(indices_[nread_++]), 0};
 }
 
+Api::SysCallSizeResult FuzzedHeader::read(void* buffer, size_t length, bool peek) {
+  const size_t len = std::min(size(), length); // Number of bytes to write
+  memcpy(buffer, data_.data() + index_, len);
+
+  if (!peek) {
+    // If not peeking, written bytes will be marked as read
+    index_ += len;
+  }
+
+  return Api::SysCallSizeResult{static_cast<ssize_t>(len), 0};
+}
+
+size_t FuzzedHeader::size() { return indices_[std::min(nread_, nreads_ - 1)] - index_ + 1; }
```

Comment:
this is saying the size is: the end index of the current read minus the start? 

is the min there because the current read can equal `nreads_`? even with the change to not increment if `done()`?